### PR TITLE
Interprocedural dead code elimination

### PR DESCRIPTION
**Pull request by Anabra**

### CODE REVIEW

File: grin/src/Grin/TypeCheck.hs [hunk 1]
```diff
@@ -31,62 +34,69 @@ import Data.Bifunctor
...
-    T_Unit    -> TypeEnv.T_Unit
-    T_Location l -> TypeEnv.T_Location [l]
+    T_Int64               -> pure TypeEnv.T_Int64
+    T_Word64              -> pure TypeEnv.T_Word64
+    T_Float               -> pure TypeEnv.T_Float
+    T_Bool                -> pure TypeEnv.T_Bool
+    T_Unit                -> pure TypeEnv.T_Unit
+    T_UnspecifiedLocation -> pure TypeEnv.T_UnspecifiedLocation
+    T_Location l          -> pure $ TypeEnv.T_Location [l]
+    l@Local{} -> throwError $ "Encountered analysis specific local value during type checking: " ++ show l
```
Comment (andorp) [inline grin/src/Grin/TypeCheck.hs]: Couldn't we decouple analysis specific values from results? I feel that the final HPTResult should not contain abstractions like "Local"

File: grin/src/Pipeline/Definitions.hs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    , _psSaveIdx       :: Int
+    , _psHPTProgram    :: Maybe HPTProgram
+    , _psHPTResult     :: Maybe HPTResult
+    , _psCByProgram    :: Maybe CByProgram
+    , _psCByResult     :: Maybe CByResult
+    , _psLVAProgram    :: Maybe LVAProgram
+    , _psLVAResult     :: Maybe LVAResult
+    , _psSharingResult :: Maybe SharingResult
+    , _psTypeEnv       :: Maybe TypeEnv
+    , _psTypeAnnots    :: Maybe TypeEnv
```
Comment (andorp) [inline grin/src/Pipeline/Definitions.hs]: Why do we need to maintain two type envs?

File: grin/src/Pipeline/Definitions.hs [hunk 3]
```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE TemplateHaskell, PatternSynonyms #-}
+module Pipeline.Definitions where
```
Comment (andorp) [inline grin/src/Pipeline/Definitions.hs]: I am against this change. Decoupling definitions from the only one module which uses them considered a bad practice in Haskell.

File: grin/src/Pipeline/Pipeline.hs [hunk 4]
```diff
@@ -1,20 +1,29 @@
-{-# LANGUAGE LambdaCase, RecordWildCards, TemplateHaskell, PatternSynonyms #-}
-module Pipeline.Pipeline where
+{-# LANGUAGE LambdaCase, RecordWildCards, RankNTypes #-}
+module Pipeline.Pipeline
+ ( module Pipeline.Pipeline 
+ , module Pipeline.Definitions
```
Comment (andorp) [inline grin/src/Pipeline/Pipeline.hs]: As they reexposed via the same interface.

File: grin/src/AbstractInterpretation/Sharing.hs [hunk 5]
```diff
@@ -57,19 +65,22 @@ calcNonLinearNonUpdateLocVariables exp = Set.fromList $ Map.keys $ Map.filter (>
       VarTagNode v ps -> union $ fmap seen (Var v : ps)
       _ -> Map.empty
 
-sharingCodeGen :: IR.Reg -> Exp -> CG ()
-sharingCodeGen s e = do
```
[the remark this answers is not in the document]
Comment (andorp) [reply grin/src/AbstractInterpretation/Sharing.hs]: Sharing is part of the HPT analysis.

File: grin/src/Pipeline/Definitions.hs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+data AbstractComputationStep
+  = CompileToAbstractProgram
+  | OptimiseAbstractProgram
+  | PrintAbstractProgram
+  | RunAbstractProgramPure
+  | PrintAbstractResult
+  deriving (Eq, Show)
+
+data SharingAnalysisStep
```
Comment (andorp) [inline grin/src/Pipeline/Definitions.hs]: Please remove as sharing should be part of the HPT run.

File: grin/app/GrinCLI.hs [hunk 7]
```diff
@@ -40,9 +41,14 @@ transformOpts =
...
   <|> flg CopyPropagation "cp" "Copy Propagation"
   <|> flg ConstantPropagation "cnp" "Constant Propagation"
-  <|> flg DeadProcedureElimination "dpe" "Dead Procedure Elimination"
-  <|> flg DeadParameterElimination "dae" "Dead Parameter Elimination"
+  <|> flg DeadCodeElimination "dce" "Dead Code Elimination"
+  <|> flg DeadDataElimination "dde" "Dead Data Elimination"
+  <|> flg DeadFunctionElimination "dfe" "Dead Function Elimination"
+  <|> flg DeadParameterElimination "dpe" "Dead Parameter Elimination"
   <|> flg DeadVariableElimination "dve" "Dead Variable Elimination"
+  <|> flg SimpleDeadFunctionElimination "sdfe" "Dead Procedure Elimination"
```
Comment (csabahruska) [inline grin/app/GrinCLI.hs]: Is *dead procedure elimination* or *dead function elimination* the better naming?
I prefer the Boq PhD naming, because it's our documentation so far.

File: grin/src/Transformations/Util.hs [hunk 8]
```diff
@@ -89,6 +95,37 @@ mapValsExp f = \case
   SUpdate name val  -> SUpdate name $ f val
   exp               -> exp
 
+mapValValM :: Monad m => (Val -> m Val) -> Val -> m Val
```
Comment (csabahruska) [inline grin/src/Transformations/Util.hs]: use Applicative style

File: grin/src/AbstractInterpretation/Sharing.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andorp) [reply grin/src/AbstractInterpretation/Sharing.hs]: Final resolution: Sharing will be separated, but it will be part of the Abstract machine.

File: grin/src/Pipeline/Definitions.hs (hunk 2 above)
Comment (andorp) [reply grin/src/Pipeline/Definitions.hs]: Please write a comment about its meaning

File: grin/app/GrinCLI.hs (hunk 7 above)
Comment (Anabra) [reply grin/app/GrinCLI.hs]: Dead Parameter Elimination's and Dead Procedure Elimination's abbreviation would clash.

File: grin/app/GrinCLI.hs [hunk 9]
```diff
@@ -106,19 +135,10 @@ main = do
         opts = defaultOpts { _poOutputDir = outputDir, _poFailOnLint = True }
     case steps of
       [] -> void $ optimize opts program prePipeline postPipeline
-      _  -> void $ pipeline opts program steps
+      _  -> void $ pipeline opts (Just content) program steps
```
Comment (andorp) [inline grin/app/GrinCLI.hs]: What is the content for?

File: grin/src/AbstractInterpretation/CByUtil.hs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+undirectedReflexiveClosure :: ProducerGraph' -> ProducerGraph'
+undirectedReflexiveClosure m = flip execState m $ do
+  let pList = Map.toList
+            . Map.map Map.toList
+            . Map.map (Map.map Set.toList)
+            $ m
+  -- for each (p, (t, [p1 .. pn])),
+  -- it add the entries: (p1, (t, [p])) .. (pn, (t, [p]))
+  -- also insert p into (p, (t, [p1 .. pn])),
+  forM pList $ \(p, taggedGroups) ->
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CByUtil.hs]: This feels like very imperative style. You could generate the lazy list and single mapM on that.

File: grin/src/AbstractInterpretation/CodeGen.hs [hunk 11]
```diff
@@ -1,433 +1,180 @@
...
+import Grin.TypeEnvDefs
 import qualified AbstractInterpretation.IR as IR
-import AbstractInterpretation.IR (HPTProgram(..), emptyHPTProgram)
+import AbstractInterpretation.IR (AbstractProgram(..), HasDataFlowInfo(..))
 
-type CG = ExceptT String (State HPTProgram)
+type CG s a = ExceptT String (State s) a
 
-data Result
+data Result s
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CodeGen.hs]: Why the 's' parameter is needed? Do we use with more than one type?

File: grin/src/AbstractInterpretation/CreatedBy.hs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                  { srcSelector = IR.ConditionAsSelector $ IR.NotIn tags
+                  , srcReg = valReg
+                  , dstReg = altScrutReg
+                  }
+            emit IR.If {condition = IR.NotIn tags, srcReg = valReg, instructions = altInstructions}
+
+          _ -> throwCBy $ "CBy does not support the following case pattern: " ++ show cpat
+
+      -- restore scrutinee register mapping
+      maybe (pure ()) (uncurry addReg) scrutRegMapping
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CreatedBy.hs]: You could use `forM_` instead of `maybe (pure ())`.

File: grin/src/Grin/Parse.hs [hunk 13]
```diff
@@ -1,140 +1,7 @@
-{-# LANGUAGE TupleSections, LambdaCase, OverloadedStrings #-}
+module Grin.Parse 
+  ( module Grin.ParseAST 
+  , module Grin.ParseTypeEnv
```
Comment (andorp) [inline grin/src/Grin/Parse.hs]: This could be in the a separate directory. `Grin\Parse\AST.hs` etc

File: grin/src/AbstractInterpretation/IR.hs [hunk 14]
```diff
@@ -64,37 +107,59 @@ data Instruction
...
-data HPTProgram
-  = HPTProgram
-  { hptMemoryCounter    :: Word32
-  , hptRegisterCounter  :: Word32
-  , hptRegisterMap      :: Map.Map Name Reg
-  , hptInstructions     :: [Instruction]
-  , hptFunctionArgMap   :: Map.Map Name (Reg, [Reg])
-  , hptTagMap           :: Bimap.Bimap Grin.Tag Tag
-  , hptSharingReg       :: Maybe Reg
+class HasDataFlowInfo a where
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/IR.hs]: Done.

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 15]
```diff
@@ -40,15 +35,22 @@ data Computer
   }
   deriving (Eq, Show)
 
-concat <$> mapM makeLenses [''NodeSet, ''Value, ''Computer]
+data ComputationResult
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: Done.

File: grin/src/Pipeline/Definitions.hs (hunk 6 above)
Comment (Anabra) [reply grin/src/Pipeline/Definitions.hs]: Now the sharing analysis will be a separate data-flow analysis. Similarly to created-by, it extends the heap-points-to analysis.

File: grin/src/AbstractInterpretation/CreatedBy.hs (hunk 12 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/CreatedBy.hs]: Done.

File: grin/src/AbstractInterpretation/CByResult.hs [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+-- HPTResult with producer info
+type HPTResultP = HPTResult
+type Producer   = Int
+
+-- node with its possible producers in its first field
+type NodeP    = Node
+-- typeSet with producer info for its nodeSet
+type TypeSetP = TypeSet
+
+mapSnd :: (b -> c) -> (a, b) -> (a, c)
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/CByResult.hs]: This code was dead, deleted it.

File: grin/src/AbstractInterpretation/IR.hs [hunk 17]
```diff
@@ -64,37 +107,59 @@ data Instruction
...
+  deriving (Generic, NFData, Eq, Ord, Show)
 
 data Constant
   = CSimpleType   SimpleType
   | CHeapLocation Mem
   | CNodeType     Tag Int {-arity-}
-  | CNodeItem     Tag Int {-node item index-} Int32 {-simple type or location-}
-  deriving Show
+  | CNodeItem     Tag Int {-node item index-} Int32 {-simple type, location, or incase of Cby: producer-}
+  deriving (Generic, NFData, Eq, Ord, Show)
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/IR.hs]: Deleted Generic and NFData derivings.

File: grin/src/AbstractInterpretation/Model.hs [hunk 18]
```diff
@@ -130,7 +130,7 @@ computeResult (BuildState maxLoc equations functions nonLinearVars warnings) =
       in result2
 
     oneStepSharing :: HPTResult -> HPTResult
-    oneStepSharing result@(HPTResult _ _ _ sharing) = result
+    oneStepSharing result@(HPTResult _ _ _ sharing) = result
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Model.hs]: Will pay attention to that.

File: grin/src/Grin/TypeCheck.hs (hunk 1 above)
Comment (Anabra) [reply grin/src/Grin/TypeCheck.hs]: Local abstract values are analysis specific artifacts. They can be used to give easy-to-understand names to these artifacts, thus differentiating them from the generic values. Also, local values should not escape the scope of the given analysis. They should be interpreted by the analysis, and transformed away.

File: grin/src/Pipeline/Definitions.hs (hunk 2 above)
Comment (Anabra) [reply grin/src/Pipeline/Definitions.hs]: Done

File: grin/src/Pipeline/Definitions.hs (hunk 3 above)
Comment (Anabra) [reply grin/src/Pipeline/Definitions.hs]: The Pipeline module is already way too big. Extracting the definitions and utilities out not only reduces the size of the module but also makes it more comprehensible. The data and the functionality are separated this way. Furthermore, the definitions are so simple, they can be understood by reading them only once. Every other occasion the programmer has to scroll through those lines is unnecessary overhead (also personal experience).

File: grin/src/Transformations/Util.hs (hunk 8 above)
Comment (Anabra) [reply grin/src/Transformations/Util.hs]: Done

File: grin/app/GrinCLI.hs (hunk 9 above)
Comment (Anabra) [reply grin/app/GrinCLI.hs]: It is used by the type annotation parser.

File: grin/src/Grin/Parse.hs (hunk 13 above)
Comment (Anabra) [reply grin/src/Grin/Parse.hs]: Done

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 19]
```diff
@@ -68,125 +70,237 @@ regIndex (Reg i) = fromIntegral i
...
+      let restrictedSrcNS = NodeSet $ Map.withoutKeys srcTagMap tagSet
+      in  mappend restrictedSrcNS dstNS
+    ValueIn rng ->
+      let filteredSrcNS = NodeSet $ Map.map (V.map (Set.filter (`inRange` rng))) srcTagMap
+      in mappend filteredSrcNS dstNS
+    ValueNotIn rng ->
+      let filteredSrcNS = NodeSet $ Map.map (V.map (Set.filter (`notInRange` rng))) srcTagMap
+      in mappend filteredSrcNS dstNS
+
+-- NOTE: a ~ Tag
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: The code is a bit more general this way.

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 20]
```diff
@@ -68,125 +70,237 @@ regIndex (Reg i) = fromIntegral i
...
       NotIn tags -> do
-        tagMap <- use $ register.ix (regIndex srcReg).nodeSet.nodeTagMap
-        typeSet <- use $ register.ix (regIndex srcReg).simpleType
-        pure $ not (Set.null typeSet) || Data.Foldable.any (flip Set.notMember tags) (Map.keysSet tagMap)
+        tagMap <- use $ selectTagMap srcReg
+        typeSet <- use $ selectReg srcReg.simpleType
+        pure $ not (Set.null typeSet) || Data.Foldable.any (`Set.notMember` tags) (Map.keysSet tagMap)
+      All (TagIn tagSet) -> do
+        tagMap <- use $ selectTagMap srcReg
+        pure $ tagPredicateCondition all Set.member tagMap
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: In my opinion, applicative style should only be used when the expressions are short, or the left-hand side is a data constructor. Also, this way the code nicely fits onto one console screen.

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 21]
```diff
@@ -68,125 +70,237 @@ regIndex (Reg i) = fromIntegral i
...
+      tagMap <- use $ selectTagMap srcReg
+      -- the union of the value sets of all fields
+      let mergedFields = mconcat . (map Data.Foldable.fold) . Map.elems $ tagMap
+      selectReg dstReg.simpleType %= (mappend mergedFields)
 
-  Move {..} -> do
-    value <- use $ register.ix (regIndex srcReg)
-    register.ix (regIndex dstReg) %= (mappend value)
+  Extend {..} -> do
+    -- TODO: support all selectors
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: This TODO item is from the original repository's master branch.

File: grin/src/AbstractInterpretation/Util.hs [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+
+reverseProgram :: HasDataFlowInfo s => s -> s
+reverseProgram = modifyInfo $ \p@AbstractProgram{..} ->
+  p { _absInstructions = reverseInstructions _absInstructions }
+
+converge :: (a -> a -> Bool) -> (a -> a) -> a -> a
+converge pred f x
+  | pred x x' = x
+  | otherwise = converge pred f x'
+  where x' = f x
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Util.hs]: Done

File: grin/src/AbstractInterpretation/Sharing.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Sharing.hs]: Done

File: grin/src/Grin/ParseBasic.hs [hunk 23]
```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Grin.ParseBasic where
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/Grin/ParseBasic.hs]: Done

File: grin/src/Pipeline/Pipeline.hs (hunk 4 above)
Comment (Anabra) [reply grin/src/Pipeline/Pipeline.hs]: See above comment.

File: grin/src/Pipeline/Optimizations.hs [hunk 24]
```diff
@@ -39,7 +39,7 @@ import Transformations.Optimising.SimpleDeadParameterElimination (simpleDeadPara
 import Transformations.Optimising.CSE (commonSubExpressionElimination)
 import Transformations.Optimising.CaseCopyPropagation (caseCopyPropagation)
 import Transformations.Optimising.GeneralizedUnboxing (generalizedUnboxing)
-import Transformations.Optimising.ArityRaisingSimple (arityRaising)
+import Transformations.Optimising.ArityRaising (arityRaising)
```
Comment (andorp) [inline grin/src/Pipeline/Optimizations.hs]: Please revert this change. The new arity raising implementation should be supported. It is well tested via the idris frontend. The test was deliberately was turned off as it is too specific for the implementation of the old arity raising implementation.

File: grin/src/Pipeline/Pipeline.hs [hunk 25]
```diff
@@ -346,6 +346,7 @@ saveTransformationInfo name content = do
   outputDir <- view poOutputDir
   let fname = printf "%03d.%s" n name
   liftIO $ do
+    createDirectoryIfMissing True outputDir
```
Comment (andorp) [inline grin/src/Pipeline/Pipeline.hs]: This is now moved into the runPipeline: https://github.com/grin-tech/grin/blob/master/grin/src/Pipeline/Pipeline.hs#L574

File: grin/src/Pipeline/Optimizations.hs (hunk 24 above)
Comment (Anabra) [reply grin/src/Pipeline/Optimizations.hs]: Done

File: grin/src/Pipeline/Pipeline.hs (hunk 25 above)
Comment (Anabra) [reply grin/src/Pipeline/Pipeline.hs]: Since `SaveGrin` can be called from anywhere,  I think we have to make sure it cannot fail due to missing directory exceptions. `runPipeline` is a really big global function, and the user might not want to call it simply to save the result of a transformation. 

As an example, the optimization pass (`--optimize`) needs to save the resulting code at each step. It uses the `SaveGrin` pipelinestep for this.

Comment (andorp) [reply grin/src/Pipeline/Pipeline.hs]: The runPipeline should be the only entry point to evaluate any Pipeline computation, thus SaveGrin shouldn't be called anywhere else just within a created pipeline computation.

The runPipeline funcion isn't that big and its responsibility to set up the computational context appropriately.
